File: src/LeetCode/P95.go
/**
@auther: kid1999
@date: 2021/8/3 9:15
@desc: P95
**/

package main

func generateTrees(n int) []*TreeNode {
	res := []*TreeNode{}
	vis := make([]bool, n+1)
	for i := 1; i <= n; i++ {
		vis[i] = true
		head := &TreeNode{i,nil,nil}
		dfs1(&res,vis,head,head,1)
		vis[i] = false
	}
	return res
}

func dfs1(res *[]*TreeNode,vis []bool,root *TreeNode,header *TreeNode,count int) {
	if count == len(vis) {
		*res = append(*res,header)
	}
	for i := 1; i < len(vis); i++{
		if !vis[i]{
			vis[i] = true
			if root.Val > i{
				root.Left = &TreeNode{i,nil,nil}
				dfs1(res,vis,root.Left,header,count+1)
			}else if root.Val < i{
				root.Right = &TreeNode{i,nil,nil}
				dfs1(res,vis,root.Right,header,count+1)
			}
			vis[i] = false
		}
	}
}

File: src/LeetCode/P95.java
package LeetCode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author kid1999
 * @create 2021-08-03 10:43
 * @description TODO
 **/
public class P95 {

    public static void main(String[] args) {
        generateTrees(3);
    }

    static List<TreeNode> res;
    static boolean[] vis;
    public static List<TreeNode> generateTrees(int n) {
        res = new ArrayList<>();
        vis = new boolean[n+1];
        for (int i = 1; i <=n ; i++) {
            vis[i] = true;
            TreeNode head = new TreeNode(i);
            dfs(head,head,1);
            vis[i] = false;
        }
        return res;
    }

    public static void dfs(TreeNode root,TreeNode head,int count){
        if (count == vis.length -1){
            res.add(head);
            return;
        }

        for (int i = 1; i < vis.length; i++) {
            if (!vis[i]){
                vis[i] = true;
                if (root.val > i){
                    root.left = new TreeNode(i);
                    dfs(root.left,head,count+1);
                }else{
                    root.right = new TreeNode(i);
                    dfs(root.right,head,count+1);
                }
                vis[i] = false;
            }
        }
    }
}